### DISCUSSION TITLE
Does using multiple processes for a python app use memory sharing?

### DISCUSSION THREAD

**Discussion by bunny-therapist** [nginx/unit#1243] [Support] [answered]
I know that for uwsgi, according to [docs](https://uwsgi-docs.readthedocs.io/en/latest/ThingsToKnow.html): 
> uWSGI tries to (ab)use the Copy On Write semantics of the [fork()](http://en.wikipedia.org/wiki/Fork_%28operating_system%29) call whenever possible. By default it will fork after having loaded your applications to share as much of their memory as possible. If this behavior is undesirable for some reason, use the lazy-apps option. This will instruct uWSGI to load the applications after each worker’s fork(). Beware as there is an older options named lazy that is way more invasive and highly discouraged (it is still here only for backward compatibility)

Does NGINX Unit do something similar? If I set `applications.python-app.processes` to, say, 5, will those 5 processes "share as much memory as possible"? Does NGINX Unit start multiple python processes, each with a copy of the app, or does it start one and then fork?

**Comment by ac000** [comment]
In short yes.

Lets look at a Unit process tree comprising unit and 5 python application processes...

```
$ ps -efH
UID          PID    PPID  C STIME TTY          TIME CMD
...
andrew      7153    7152  0 15:11 pts/2    00:00:00         -bash
andrew      7184    7153  0 15:11 pts/2    00:00:00           unit: main v1.33.0 [/opt/unit/sbin/unitd --no-daemon]
andrew      7186    7184  0 15:11 pts/2    00:00:00             unit: controller
andrew      7187    7184  0 15:11 pts/2    00:00:00             unit: router
andrew      7235    7184  0 15:12 pts/2    00:00:00             unit: "python" prototype
andrew      7236    7235  0 15:12 pts/2    00:00:00               unit: "python" application
andrew      7237    7235  0 15:12 pts/2    00:00:00               unit: "python" application
andrew      7238    7235  0 15:12 pts/2    00:00:00               unit: "python" application
andrew      7239    7235  0 15:12 pts/2    00:00:00               unit: "python" application
andrew      7240    7235  0 15:12 pts/2    00:00:00               unit: "python" application
...
```

You'll notice we create a "prototype" process, we do that for each application (regardless of language), from there we fork the prescribed number of "application" processes.

You can see that the _application_ processes are _children_ of the _prototype_ process.

Prototype processes were added in e207415a78ae67b937faf7e5bcd6e5192993180a

Applications processes may come and go but the prototype process will live for the duration of the application.

**Reply by bunny-therapist** [reply]
Thank you very much!

**Reply by bunny-therapist** [reply]
Is garbage collection frozen for preexisting objects (or otherwise handled) after the fork? I am asking since I know that forking in python without doing that risk triggering copy-on-write just because of garbage collection (garbage collection being "write").

**Reply by ac000** [reply]
Not AFAIACT...

**Reply by bunny-therapist** [reply]
Not sure how often it is relevant. I originally learned about it from this video: https://www.youtube.com/watch?v=Hgw_RlCaIds
(From the video, it seems like maybe freeze should be _before_ the fork, not after it.)

**Reply by ac000** [reply, marked as answer]
Thanks for the link to the video. That made me dig a little deeper.

My answer here may change slightly what you take away from my original answer.

Now gc.freeze is some Python API and obviously when we do the fork(2) we are still in C, but that is all moot anyway...

The _prototype_ processes _only_ load the required Unit language module, e.g for python `$(modulesdir}/python.unit.so`, which in turn loads the required libraries for that language. e.g the python prototype process loads these (the PHP language module loads a lot more things...)

```
$ pmap -p PID
...
00007f0c7d800000    496K r---- /usr/lib64/libpython3.12.so.1.0
00007f0c7d87c000   2180K r-x-- /usr/lib64/libpython3.12.so.1.0
00007f0c7da9d000   1340K r---- /usr/lib64/libpython3.12.so.1.0
00007f0c7dbec000    480K r---- /usr/lib64/libpython3.12.so.1.0
00007f0c7dc64000   1460K rw--- /usr/lib64/libpython3.12.so.1.0
...
00007f0c7de04000     20K r---- /opt/unit/modules/python.unit.so
00007f0c7de09000     76K r-x-- /opt/unit/modules/python.unit.so
00007f0c7de1c000     28K r---- /opt/unit/modules/python.unit.so
00007f0c7de23000      4K r---- /opt/unit/modules/python.unit.so
00007f0c7de24000      8K rw--- /opt/unit/modules/python.unit.so
...
00007f0c7de29000    160K r---- /usr/lib64/libc.so.6
00007f0c7de51000   1444K r-x-- /usr/lib64/libc.so.6
00007f0c7dfba000    312K r---- /usr/lib64/libc.so.6
00007f0c7e008000     16K r---- /usr/lib64/libc.so.6
00007f0c7e00c000      8K rw--- /usr/lib64/libc.so.6
...
00007f0c7e016000      8K r---- /usr/lib64/libpcre2-8.so.0.11.2
00007f0c7e018000    448K r-x-- /usr/lib64/libpcre2-8.so.0.11.2
00007f0c7e088000    160K r---- /usr/lib64/libpcre2-8.so.0.11.2
00007f0c7e0b0000      4K r---- /usr/lib64/libpcre2-8.so.0.11.2
00007f0c7e0b1000      4K rw--- /usr/lib64/libpcre2-8.so.0.11.2
00007f0c7e0b2000     64K r---- /usr/lib64/libm.so.6
00007f0c7e0c2000    476K r-x-- /usr/lib64/libm.so.6
00007f0c7e139000    360K r---- /usr/lib64/libm.so.6
00007f0c7e193000      4K r---- /usr/lib64/libm.so.6
00007f0c7e194000      4K rw--- /usr/lib64/libm.so.6
...
00007f0c7e19d000      4K r---- /usr/lib64/ld-linux-x86-64.so.2
00007f0c7e19e000    156K r-x-- /usr/lib64/ld-linux-x86-64.so.2
00007f0c7e1c5000     40K r---- /usr/lib64/ld-linux-x86-64.so.2
00007f0c7e1cf000      8K r---- /usr/lib64/ld-linux-x86-64.so.2
00007f0c7e1d1000      8K rw--- /usr/lib64/ld-linux-x86-64.so.2
...
```

The _prototype_ process doesn't load or run any application code, that _only_ happens in the _application_ processes.

**Reply by bunny-therapist** [reply]
Oh, ok so then I would say the answer to my original question is more of a "no" than a "yes" (even though what you are doing there is of course very good).

For me, the issue was to do with applications which load machine learning models, and trying to run multiple copies of such an app without having multiple copies of the model in memory. I noticed that when was doing things like this with uwsgi, having multiple processes did not affect the total memory consumption significantly, but when I was using NGINX Unit, each copy of the app added 0.5 GB of memory consumption (which seems to be approximately how much memory is used by the loaded model). This means if I still want the same performance of my app, I will have to use _much_ more memory.

I see that gunicorn also has some support for this, using a "preload" option: https://www.joelsleppy.com/blog/gunicorn-application-preloading/

How would you feel about my making this a feature request?

**Reply by ac000** [reply]
It's an interesting problem/idea...

At the moment I couldn't definitively say that it is something we would want to implement or even _could_ implement.

Remember we have more than just Python to worry about. Then there is the question of if we loaded an application that opened a bunch of database connections or somesuch, then when we forked the application processes each process would have the _same_ connections and that's not gong to end well.

Then there is the _reload_ issue as mentioned which may be problematic for us.

How are these _models_ loaded? Are they contained in files that are say mmap(2)ed or are they generated at run-time?

I can see ways of doing this, but of course it requires a bit more work on the application side...

**Reply by bunny-therapist** [reply]
I will say, to me, this way of doing stuff that uwsgi does (at least if I understand the docs correctly) is a bit strange, and they even call it "(ab)use" in their docs. So I fully understand if you consider it too hacky. I can't really make any recommendations myself.

As for the models in my case, they are tar balls that are decompressed, then the tar balls contain some pickle files, which are depickled into python objects. The decompression and deserialization is entirely from python, using tarfile and pickle libraries. (I don't really think this pickling is a good format, personally.)

Sadly I don't even know what "mmap(2)" is (even after trying to look it up) so I can't really answer that part of your question.
